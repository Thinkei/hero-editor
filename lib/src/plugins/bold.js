import React from 'react';
import { useSlate } from 'slate-react';
import { isMarkActive, toggleMark, makeRenderLeaf } from '../helpers';

const renderLeaf = makeRenderLeaf('bold', ({ children }) => (
  <strong>{children}</strong>
));

const ToolbarButton = () => {
  const editor = useSlate();

  return (
    <button
      style={{
        background: isMarkActive(editor, 'bold') ? 'grey' : 'lightgrey',
      }}
      onMouseDown={(event) => {
        event.preventDefault();
        toggleMark(editor, 'bold');
      }}
    >
      bold
    </button>
  );
};

export default { renderLeaf, ToolbarButton };
