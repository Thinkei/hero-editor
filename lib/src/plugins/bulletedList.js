import React from 'react';
import { Transforms } from 'slate';
import { useSlate } from 'slate-react';
import {
  createPlugin,
  isBlockActive,
  makeRenderElement,
  addMessageListener,
  postMessage,
  isList,
} from '../helpers';
import Toolbar from '../components/Toolbar';
import Icon from '../components/Icon';
import listItem from './listItem';
import { BULLETED_LIST } from '../constants';

const renderElement = makeRenderElement(BULLETED_LIST, 'ul');

const handleMessage = addMessageListener(BULLETED_LIST, ({ editor }) => {
  const isActive = isBlockActive(editor, BULLETED_LIST);

  // Unwrap all the items of any list
  Transforms.unwrapNodes(editor, {
    match: isList,
    split: true,
  });

  if (isActive) {
    Transforms.setNodes(editor, { type: 'paragraph' });
  } else {
    Transforms.setNodes(editor, { type: listItem.name });
    Transforms.wrapNodes(editor, { type: BULLETED_LIST, children: [] });
  }
});

const ToolbarButton = () => {
  const editor = useSlate();

  return (
    <Toolbar.Button
      active={isBlockActive(editor, BULLETED_LIST)}
      onMouseDown={(event) => {
        event.preventDefault();
        postMessage(BULLETED_LIST, {}, editor);
      }}
    >
      <Icon>format_list_bulleted</Icon>
    </Toolbar.Button>
  );
};

export default createPlugin({
  name: BULLETED_LIST,
  renderElement,
  handleMessage,
  ToolbarButton,
});
