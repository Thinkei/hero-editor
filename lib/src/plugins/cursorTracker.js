import { ReactEditor } from 'slate-react';
import { createPlugin, addMessageListener, postMessage } from '../helpers';
import { EDITOR_CHANGE } from '../constants';

const handleMessage = addMessageListener(EDITOR_CHANGE, ({ editor }) => {
  const { selection } = editor;

  if (selection) {
    const domRange = ReactEditor.toDOMRange(editor, selection);
    const rangeRect = domRange.getBoundingClientRect();

    postMessage(
      'cursor-change',
      {
        offset: undefined,
        position: {
          top: rangeRect.top,
          left: rangeRect.left,
        },
      },
      editor,
    );
  }
});

export default () =>
  createPlugin({
    name: 'cursor-tracker',
    handleMessage,
  });
