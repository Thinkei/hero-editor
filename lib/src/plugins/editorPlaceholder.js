import React, { useCallback, useEffect, useRef } from 'react';
import { useSlate } from 'slate-react';
import { createPlugin, useForceUpdate } from '../helpers';

const styles = {
  placeholder: {
    boxSizing: 'border-box',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    padding: 16,
    lineHeight: 1.5,
    color: 'lightgrey',
    zIndex: -1,
  },
};

const CACHE_TTL = 500;

const isDefined = (ref) => typeof ref.current !== 'undefined';

const checkEmptyOptimistically = (content) =>
  content.length === 1 &&
  content[0].type === 'paragraph' &&
  content[0].children.length === 1 &&
  content[0].children[0].text === '';

const EditorPlaceholder = ({ value }) => {
  const editor = useSlate();
  const cache = useRef(undefined);
  const { isForceUpdated, forceUpdate } = useForceUpdate();

  const createPlaceholder = useCallback(() => {
    const isEmpty = checkEmptyOptimistically(editor.children);

    return isEmpty ? <div style={styles.placeholder}>{value}</div> : null;
  }, []);

  useEffect(() => {
    if (isDefined(cache)) {
      const timerId = setTimeout(() => {
        cache.current = undefined;

        if (!isForceUpdated) forceUpdate();
      }, CACHE_TTL);

      return () => clearTimeout(timerId);
    }
  });

  if (isDefined(cache)) return cache.current;

  cache.current = createPlaceholder();

  return cache.current;
};

export default ({ value }) =>
  createPlugin({
    name: 'editor-placeholder',
    renderCustom: () => (
      <EditorPlaceholder key="editor-placeholder" value={value} />
    ),
  });
