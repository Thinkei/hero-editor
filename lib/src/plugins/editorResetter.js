import { createPlugin, addMessageListener } from '../helpers';
import { EMPTY_VALUE } from '../constants';

const name = 'editor-resetter';

const handleMessage = addMessageListener(
  'reset',
  ({ editor, data = EMPTY_VALUE }) => {
    editor.children = data;
    editor.selection = {
      anchor: { path: [0, 0], offset: 0 },
      focus: { path: [0, 0], offset: 0 },
    };
    editor.onChange();
  },
);

export default () =>
  createPlugin({
    name,
    handleMessage,
  });
