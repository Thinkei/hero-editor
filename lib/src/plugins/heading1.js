import React from 'react';
import { useSlate } from 'slate-react';
import {
  createPlugin,
  isMarkActive,
  toggleMark,
  makeRenderLeaf,
  addMessageListener,
  postMessage,
} from '../helpers';
import Toolbar from '../components/Toolbar';
import Icon from '../components/Icon';
import { H1 } from '../constants';

const renderLeaf = makeRenderLeaf(H1, ({ children }) => <h1>{children}</h1>);

const handleMessage = addMessageListener(H1, ({ editor }) => {
  toggleMark(editor, H1);
});

const ToolbarButton = () => {
  const editor = useSlate();

  return (
    <Toolbar.Button
      active={isMarkActive(editor, H1)}
      onMouseDown={(event) => {
        event.preventDefault();
        postMessage(H1, {}, editor);
      }}
    >
      <Icon>format_h1</Icon>
    </Toolbar.Button>
  );
};

export default () =>
  createPlugin({
    name: H1,
    renderLeaf,
    handleMessage,
    ToolbarButton,
  });
