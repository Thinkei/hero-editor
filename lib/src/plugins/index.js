export { default as bold } from './bold';
export { default as italic } from './italic';
export { default as underline } from './underline';
export { default as logger } from './logger';
export { default as bulletedList } from './bulletedList';
export { default as numberedList } from './numberedList';
export { default as listItem } from './listItem';
