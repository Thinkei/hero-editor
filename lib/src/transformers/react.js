import React from 'react';
import { Text, Node } from 'slate';
import y from '../helpers/why';

const isRoot = Array.isArray;

const makeReactTransformer = (rules) =>
  y((reactTransformer) => (node, key) => {
    if (isRoot(node)) return node.map(reactTransformer);

    if (Text.isText(node)) return rules.leaf(node, key);

    const children = node.children.map(reactTransformer);

    return rules.element(node, children, key);
  });

const defaultRules = {
  leaf: (node, key) => {
    let children = node.text;

    if (node.bold) children = <strong key={key}>{children}</strong>;

    if (node.italic) children = <em key={key}>{children}</em>;

    if (node.underline) children = <u key={key}>{children}</u>;

    return <span key={key}>{children}</span>;
  },

  element: (node, children, key) => {
    switch (node.type) {
      case 'paragraph':
        return (
          <p
            key={key}
            style={{ margin: 0, lineHeight: 1.5, minHeight: '1.5em' }}
          >
            {children}
          </p>
        );

      case 'heading-one':
        return <h1 key={key}>{children}</h1>;

      case 'link':
        return (
          <a key={key} href={Node.string(node)} target="_blank">
            {children}
          </a>
        );

      case 'mention':
        return (
          <span
            key={key}
            style={{
              padding: '2px 4px',
              background: '#eeeeee',
              borderRadius: 4,
            }}
          >
            @{node.data.name}
          </span>
        );

      case 'bulleted-list':
        return <ul key={key}>{children}</ul>;

      case 'numbered-list':
        return <ol key={key}>{children}</ol>;

      case 'list-item':
        return <li key={key}>{children}</li>;

      default:
        return <div key={key}>{children}</div>;
    }
  },
};

const defaultReactTransformer = makeReactTransformer(defaultRules);

export default makeReactTransformer;

export { makeReactTransformer, defaultReactTransformer };
