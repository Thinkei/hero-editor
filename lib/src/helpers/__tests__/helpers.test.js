import React from 'react';
import { render } from '@testing-library/react';
import { compose, flow, composeRenderLeaf, makeRenderLeaf } from '../index';
import { bold, italic, underline } from '../../plugins';

describe('compose', () => {
  it('composes functions', () => {
    expect(
      compose(
        (x) => x + 1,
        undefined,
        (x) => x * 2,
      )(10),
    ).toEqual(21);
  });
});

describe('flow', () => {
  it('... composes functions, too', () => {
    expect(
      flow(
        (x) => x + 1,
        undefined,
        (x) => x * 2,
      )(10),
    ).toEqual(22);
  });
});

describe('makeRenderLeaf', () => {
  it('does not block the next render', () => {
    const renderLeaf = makeRenderLeaf('foo', 'code');
    const next = jest.fn();
    renderLeaf(next)({
      leaf: { foo: true },
      children: 'baz',
    });
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      leaf: { foo: true },
      children: <code leaf={{ foo: true }}>baz</code>,
    });
  });
});

describe('composeRenderLeaf', () => {
  it('renders nested leaf nodes in correct order', () => {
    const renderLeaf = composeRenderLeaf([bold(), italic(), underline()]);
    const { baseElement } = render(
      renderLeaf({
        leaf: { bold: true, italic: true, underline: true },
        children: 'Rich text',
      }),
    );
    expect(baseElement).toMatchSnapshot();
  });
});
