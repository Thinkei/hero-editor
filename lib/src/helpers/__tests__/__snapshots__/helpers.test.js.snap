// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeRenderLeaf renders nested leaf nodes in correct order 1`] = `
<body>
  <div>
    <span>
      <u>
        <em>
          <strong>
            Rich text
          </strong>
        </em>
      </u>
    </span>
  </div>
</body>
`;
